### PR TITLE
Add new prompt and wrap gp command

### PR DESCRIPTION
**Pull request by hanseartic**
* Wrapping the calls to `gp` command inside it's own function allows for more graceful handling when running locally
* A new ` browse` function tries to open a web-browser environment-agnostic

<a href="https://gitpod.io/#https://github.com/tyvdh/soroban-quest--pioneer/pull/22"></a>

### CODE REVIEW

File: _squirtle/mod.ts [hunk 1]
```diff
@@ -437,45 +473,115 @@ const runRPC = async (argv: any) => {
...
+}
+
 const runHelp = async () => {
   const run1 = Deno.run({
     cmd: ['sq', 'help'],
   })
   await run1.status()
 }
 
+const gp = async (opts: Deno.RunOptions): Promise<string> => {
```
Comment (kalepail) [inline _squirtle/mod.ts]: Screen Shot 2022-11-29 at 9 16 00 PM
https://eslint.org/docs/latest/rules/no-async-promise-executor

File: _squirtle/bash-rpc-prompt [hunk 2]
```diff
@@ -1 +1 @@
-export PS1="\$(sq rpc -s) ${PS1}"
\ No newline at end of file
+export PS1="\$(sq rpc -p) ${PS1}"
```
Comment (kalepail) [inline _squirtle/bash-rpc-prompt]: Is this firing for every invocation in the futurenet bash? I'm starting to think I may not like this. Need to play around with it a bit more.

File: _squirtle/mod.ts [hunk 3]
```diff
@@ -46,22 +44,21 @@ const runLogin = async () => {
     discordUrl.searchParams.append('prompt', 'consent');
     discordUrl.searchParams.append('state', gitpodUrl.toString());
 
-    const run2 = Deno.run({
-      cmd: ['gp', 'preview', '--external', discordUrl.toString()]
-    })
-    await run2.status()
+    await gp({
```
Comment (kalepail) [inline _squirtle/mod.ts]: Why not use browse here?

File: _squirtle/mod.ts [hunk 4]
```diff
@@ -70,29 +67,82 @@ const runLogin = async () => {
   await runUser(null, user, env)
 }
 
+const browse = (location: string): Promise<Deno.ProcessStatus> => {
+  return new Promise<Deno.ProcessStatus>((resolve, reject) => {
+    switch (process.platform) {
+      case 'aix':
+      case 'freebsd':
+      case 'linux':
```
Comment (kalepail) [inline _squirtle/mod.ts]: Gitpod is always linux right? Why include other processes?

File: _squirtle/mod.ts [hunk 5]
```diff
@@ -70,29 +67,82 @@ const runLogin = async () => {
...
+          stdout: 'null',
+        })
+          .status()
+          .then(resolve)
+          .catch(reject)
+    }
+  })
+    .then((status: Deno.ProcessStatus) => {
+      if (status.success)
+        console.log("opened %s", location)
```
Comment (kalepail) [inline _squirtle/mod.ts]: I don't think I want to log every time an external site opens. The site opens, that's the notification. Right?

File: _squirtle/mod.ts [hunk 6]
```diff
@@ -70,29 +67,82 @@ const runLogin = async () => {
...
 const runLogout = async (
-  _: any, 
+  _: any,
   internal = false
 ) => {
-  const run1 = Deno.run({
-    cmd: ['gp', 'env', '-u', 'AUTH_TOKEN'],
+  const { AUTH_TOKEN, ENV } = await getEnv()
+  if (!AUTH_TOKEN)
+    return console.log(`Please run the <login> command first`);
```
Comment (kalepail) [inline _squirtle/mod.ts]: ?? Please login before you logout? Let's just skip this and assume someone is already logged out.

File: _squirtle/mod.ts [hunk 7]
```diff
@@ -70,29 +67,82 @@ const runLogin = async () => {
...
-    run1.status(),
-    run2.status(),
-    run3.status(),
-    run4.status(),
-  ])
+    run1,
+    run2,
+    run3,
+    run4,
+  ]).catch()
```
Comment (kalepail) [inline _squirtle/mod.ts]: Does this work? Can you call catch without a function arg? Might be better to just let the Deno execution catch or else if you really want to ignore maybe `.catch(() => {})` but if this has been tested and works 👍

File: _squirtle/mod.ts [hunk 8]
```diff
@@ -107,12 +157,12 @@ const runUser = async (
...
 
   const { AUTH_TOKEN, ENV } = env
-  const isDev = ENV !== 'prod'
-  const siteUrl = isDev ? 'https://quest-dev.stellar.org' : 'https://quest.stellar.org'
-
   if (!AUTH_TOKEN)
     return console.log(`Please run the <login> command first`);
 
+  const isDev = ENV !== 'prod'
+  const siteUrl = isDev ? 'https://quest-dev.stellar.org' : 'https://quest.stellar.org'
```
Comment (kalepail) [inline _squirtle/mod.ts]: Makes more sense to me to keep these at the top of the function but if you are going to move them just move them all the way down into the if block where we actually use `siteUrl`

File: _squirtle/mod.ts [hunk 9]
```diff
@@ -257,25 +302,24 @@ const runCheck = async (argv: any) => {
...
    Would you like to connect your Stellar wallet?`);
 
     if (missingPkConfirmed) {
-      const run1 = Deno.run({
-        cmd: ['gp', 'preview', '--external', `${siteUrl}/settings`]
-      })
-      return run1.status()
+      return gp({
+        cmd: ['preview', '--external', `${siteUrl}/settings`]
+      }).catch(() => { console.warn('❌ This only works inside a gitpod') })
```
Comment (kalepail) [inline _squirtle/mod.ts]: Again why not use `browse` and why the catch? `sq` is meant for Gitpod only I don't want to pepper non-gitpod related errors everywhere.

File: _squirtle/mod.ts [hunk 10]
```diff
@@ -300,11 +341,10 @@ const runCheck = async (argv: any) => {
   if (!claimToken) // No claim token but also no error, you've already solved
     return console.log("🎉 Correct! 🧠");
 
-  const run3 = Deno.run({
-    cmd: ['gp', 'env', `CLAIM_TOKEN=${claimToken}`],
-    stdout: 'null'
+  await gp({
+    cmd: ['env', `CLAIM_TOKEN=${claimToken}`],
+    stdout: 'null', stderr: 'null'
```
Comment (kalepail) [inline _squirtle/mod.ts]: Isn't null the default? If it's not should it be? Then we don't need to be explicit here.

File: _squirtle/mod.ts [hunk 11]
```diff
@@ -406,11 +437,23 @@ const runSubmit = async (argv: any) => {
 }
 
 const runRPC = async (argv: any) => {
```
Comment (kalepail) [inline _squirtle/mod.ts]: I think I'd like all this rpc stuff split out into it's own PR. It's fine if it's dependent on this PR getting merged but this is an entirely separate interest than the rest of this PR and should be able to be reviewed independent of the other tooling and plumbing tweaks

File: _squirtle/mod.ts [hunk 12]
```diff
@@ -437,45 +480,115 @@ const runRPC = async (argv: any) => {
...
+  }).catch(() => { })
 }
 
 const getEnv = async () => {
-  const run1 = Deno.run({
-    cmd: ['gp', 'env'],
-    stdout: 'piped'
-  })
-  const gpEnvString = new TextDecoder().decode(await run1.output()).trim()
+  const gpEnvString = await gp({ cmd: ['env'] })
```
Comment (kalepail) [inline _squirtle/mod.ts]: This should still be piped right otherwise we get no env string?

File: _squirtle/mod.ts [hunk 13]
```diff
@@ -577,43 +700,72 @@ const doFund = (pk: string) => {
...
 
-const getRPCStatus = () => {
-  return fetch('http://127.0.0.1:8000')
+type rpcStatusCode = 'unknown' | 'booting' | 'catching_up' | 'ready'
+interface RPCStatus {
+  ready: boolean,
+  status: rpcStatusCode,
+}
+
+const getRPCStatus = async (horizon: URL): Promise<RPCStatus> => {
```
Comment (kalepail) [inline _squirtle/mod.ts]: Great changes but again let's pull out into that second PR

File: _squirtle/mod.ts [hunk 14]
```diff
@@ -577,43 +700,72 @@ const doFund = (pk: string) => {
...
+enum knownHorizons {
+  KanayeNet = "https://kanaye-futurenet.stellar.quest:443",
+  nebolsin = "https://nebolsin-futurenet.stellar.quest:443",
+  kalepail = "https://kalepail-futurenet.stellar.quest:443",
+  silence = "https://silence-futurenet.stellar.quest:443",
+  Raph = "https://raph-futurenet.stellar.quest:443",
+  nesho = "https://nesho-futurenet.stellar.quest:443",
 }
 
 const selectRPCEndpoint = async () => {
```
Comment (kalepail) [inline _squirtle/mod.ts]: For second PR

File: _squirtle/mod.ts [hunk 15]
```diff
@@ -566,9 +679,19 @@ const autoFund = async (pk: string) => {
     return doFund(pk)
 }
 
+const LOCAL_HORIZON = 'http://127.0.0.1:8000'
+const SOROBAN_RPC_URI = '/soroban/rpc'
```
Comment (kalepail) [inline _squirtle/mod.ts]: I like this however
A) these variables should be used universally
B) I want to extend this out to the rest of the site so anywhere we're making calls to horizon we're using the currently selected endpoint vs defaulting to the local one (see https://discord.com/channels/763798356484161566/772838189920026635/1044288768032649337)

File: _squirtle/bash-rpc-prompt (hunk 2 above)
Comment (kalepail) [reply _squirtle/bash-rpc-prompt]: Also let's move this to the second PR (see below comments for splitting this PR into two PRs)

**Review by kalepail — CHANGES_REQUESTED** [review]
Overall a great PR with lots of improvements. I do have a number of comments, requests and questions and ultimately will want this split into two PRs. One with the plumbing and tooling changes (the focus of this review so keep those changes in this PR) and a second with all your Horizon/RPC fixes and improvements. (which I will review separately in that PR)

Thanks!

File: _squirtle/mod.ts (hunk 12 above)
Comment (hanseartic) [reply _squirtle/mod.ts]: `gp` is always doing the pipe internally by default

File: _squirtle/mod.ts (hunk 3 above)
Comment (hanseartic) [reply _squirtle/mod.ts]: just forgot to. well spotted

File: _squirtle/mod.ts (hunk 6 above)
Comment (hanseartic) [reply _squirtle/mod.ts]: it was there before - did not remove it. but it makes perfect sense to do so

File: _squirtle/mod.ts (hunk 15 above)
Comment (hanseartic) [reply _squirtle/mod.ts]: actually calls are made to the currently selected - the EP is resolved via getHorizonEndpoint()` before the call. this const is the fallback

Comment (kalepail) [reply _squirtle/mod.ts]: Ah okay as long as anywhere we were calling http://127.0.0.1:8000 hard coded we aren't anymore and are using whatever has been set by `sq rpc -c`

File: _squirtle/mod.ts (hunk 12 above)
Comment (kalepail) [reply _squirtle/mod.ts]: 👍 as long as you're confident and it's tested I'm cool with it

File: _squirtle/mod.ts (hunk 15 above)
Comment (hanseartic) [reply _squirtle/mod.ts]: that's what this is supposed to support, yes

**Comment by hanseartic** [comment]
Thx for the feedback. I removed the RPC related stuff and will open another PR #23
